### ISSUE TITLE
how can embedding_bias be used?

### ISSUE THREAD

**Issue by tonylek** [NVIDIA/TensorRT-LLM#1457]
Hi,
I've seen there is an option to add embedding_bias to my requests.
Is there any documentation how to work with it and what exactly it does?
I'm especially interested if this is the relevant feature for the use-case where I want to add some bias to my decoding process.
For example if in the first token I don't want any possible token but just from a closed list of tokens.

**Comment by nekorobov** [comment]
Hi @tonylek , documentation about `embedding_bias` is indeed missing. We'll add it, thank you for raising this question.

The `embedding_bias` is a vector consisting of floating-point numbers, with a length equal to `[vocabSize]`. This bias is added to the logits for each token in the vocabulary before decoding occurs. Positive values in the bias encourage the sampling of tokens, while negative values discourage it. A value of `0.f` leaves the logit value unchanged. 

`embedding_bias` is set once per request's lifetime at `SamplingConfig` and can't be changed during the sampling depending on the timestep. I believe, for your use-case you need `LogitsPostProcessor`. See documentation [here](https://github.com/NVIDIA/TensorRT-LLM/blob/main/docs/source/batch_manager.md#logits-post-processor-optional). With this callback you can manually modify logits depending on the timestep before they are passed to the decoder. Simply set `-INF` to the logits of tokens that you want to ban from being sampled.

**Comment by tonylek** [comment]
The Logits Post processor seems like the solution but I'm not so sure how can I send it or modify it. I see this is a cpp code. Where can I modify it? How does it integrate with triton?

**Comment by MartinMarciniszyn** [comment]
Unfortunately, this is not supported from Triton at the moment.

**Comment by tonylek** [comment]
What alternatives for triton exist for serving trtllm models?

**Comment by nekorobov** [comment]
You can try to customly extend Triton server code to support it. Modify [here ](https://github.com/triton-inference-server/tensorrtllm_backend/blob/bf5e9007a3f16c7fc76cb156a3362d1caae198dd/inflight_batcher_llm/src/model_instance_state.cc#L584) and [here ](https://github.com/triton-inference-server/tensorrtllm_backend/blob/bf5e9007a3f16c7fc76cb156a3362d1caae198dd/inflight_batcher_llm/src/model_instance_state.cc#L612) to call `setLogitsPostProcessor` to provide the callback. I hope it helps.

**Comment by tonylek** [comment]
This is super helpful. If somewhere there is an example for custom implementation of setLogitsPostProcessor this will be super helpful since I'm not 100% sure how to work with the arguments this method gets.
Thanks a lot!

**Comment by nekorobov** [comment]
`setLogitsPostProcessor` is simply a setter of the callback of format 
`using LogitsPostProcessor = std::function<TensorPtr(RequestIdType, TensorPtr&, BeamTokens const&, TStream)>;` to the request that is then passed to the runtime. I afraid we don't have any public example of this callback. 

I think, smth like this is what you need:
```
    LlmRequest::LogitsPostProcessor logitsCb
        = [&step](uint64_t rId, tensorrt_llm::runtime::ITensor::SharedPtr& logits,
              const LlmRequest::BeamTokens& tokens, tensorrt_llm::runtime::BufferManager::CudaStreamPtr streamPtr)
    {
        auto tensorAten = tensorrt_llm::runtime::Torch::tensor(logits);
        auto bias = at::full_like(tensorAten, c10::Scalar(-SOME_VAL), {at::kGPU});
        
        // apply some logic to mask according to the step of the request.
        
        auto atStream = tensorrt_llm::runtime::TorchUtils::stream(*streamPtr);

        {
            at::cuda::CUDAStreamGuard guard(atStream);
            tensorAten += mask.to({at::kCUDA}, true);
        }
        
        step += 1;
        
        return logits;
    };

```
